devenv/main.py

```diff
@@ -67,47 +33,49 @@ def devenv(argv: Sequence[str], config_path: str) -> ExitCode:
     config = read_config(config_path)
 
     # Guessing temporary code root
-    coderoot = config.get("devenv", "coderoot", fallback=None) or (
+    code_root = config.get("devenv", "coderoot", fallback=None) or (
         os.path.abspath(f"{current_root}/..")
         if current_root
         else os.path.expanduser("~/code")
     )
 
-    if args.command == "bootstrap":
-        return bootstrap.main(config_path, coderoot, remainder)
+    modinfo_list: Sequence[DevModuleInfo] = [
+        module.module_info
+        for module in [bootstrap, fetch, doctor, pin_gha, sync]
+        if hasattr(module, "module_info")
+    ]
 
-    if args.command == "fetch":
-        return fetch.main(coderoot, remainder)
+    # TODO: Search for modules in work repo
 
-    if not args.nocoderoot and not os.getcwd().startswith(coderoot):
-        print(
-            f"You aren't in your code root ({coderoot})!\n"
-            "To ignore, use devenv --nocoderoot [COMMAND]\n"
-            f"To change your code root, you can edit {config_path}.\n"
-        )
-        return 1
+    parser = argparse.ArgumentParser()
+    subparser = parser.add_subparsers(
+        title=argparse.SUPPRESS,
+        metavar="command",
+        dest="command",
+        required=True,
+    )
 
-    # the remaining tools are repo-specific
-    reporoot = gitroot()
+    for info in modinfo_list:
+        # Argparse stuff
+        subparser.add_parser(info.command, help=info.help)
 
-    fake_reporoot = os.getenv("CI_DEVENV_INTEGRATION_FAKE_REPOROOT")
-    if fake_reporoot:
-        reporoot = fake_reporoot
+    args, remainder = parser.parse_known_args(argv[1:])
 
-    repo = reporoot.split("/")[-1]
-    context = {"repo": repo, "reporoot": reporoot}
+    # context for subcommands
+    context: Context = {
+        "config_path": config_path,
+        "code_root": code_root,
+        "repo": Repository(current_root) if current_root else None,
+    }
 
-    if args.command == "doctor":
-        return doctor.main(context, remainder)
-    if args.command == "sync":
-        return sync.main(context, remainder)
-
-    return 1
+    command_actions = {info.command: info.action for info in modinfo_list}
+    action = command_actions.get(args.command)
+    assert action is not None
+    return action(context, remainder)
```

Comment:
Played around with this a bit and found it worked quite well. One area that could use improvement is the help messages which give very little if any useful information (for specific actions). This doesn't have to be addressed in this PR, but something to keep an eye on.